src/ui/public/visualize/visualization.js

```diff
@@ -17,137 +17,71 @@
-        const render$ = Rx.Observable.create(observer => {
-          $scope.$on('render', () => {
-            observer.next({
-              vis: $scope.vis,
-              visData: $scope.visData,
-              container: getVisContainer(),
-            });
-          });
-        }).pipe(
-          share()
-        );
-
-        const success$ = render$.pipe(
-          tap(() => {
-            dispatchRenderStart($el[0]);
-          }),
-          filter(({ vis, visData, container }) => vis && vis.initialized && container && (!vis.type.requiresSearch || visData)),
-          debounceTime(100),
-          switchMap(async ({ vis, visData, container }) => {
-            vis.size = [container.width(), container.height()];
-            const status = getUpdateStatus(vis.type.requiresUpdateStatus, $scope);
-            const renderPromise = visualization.render(visData, status);
-            $scope.$apply();
-            return renderPromise;
-          })
-        );
-
-        const requestError$ = render$.pipe(
-          filter(({ vis }) => vis.requestError)
-        );
-
-        const renderSubscription = Rx.merge(success$, requestError$)
-          .subscribe(() => {
-            dispatchRenderComplete($el[0]);
-          });
-
-        $scope.$on('$destroy', () => {
-          visualization.destroy();
-          renderSubscription.unsubscribe();
-        });
-
-        // Listen on changes to trigger a render if listenOnChange is not false
-        // i.e. true or has not been used (undefined in that case)
-        if ($scope.listenOnChange !== false) {
-          const onChangeListener = () => {
-            $scope.$emit('render');
-          };
-
-          $scope.$watchGroup(['visData', 'vis.params'], onChangeListener);
-
-          const resizeChecker = new ResizeChecker($el);
-          resizeChecker.on('resize', () => {
-            $scope.$emit('render');
-          });
-
-          $scope.uiState.on('change', onChangeListener);
-          $scope.$on('$destroy', () => {
-            resizeChecker.destroy();
-            $scope.uiState.off('change', onChangeListener);
-          });
-        }
-
-        function jQueryGetter(selector) {
-          return function () {
-            const $sel = $el.find(selector);
-            if ($sel.length) return $sel;
-          };
-        }
-      }
+    this.state = {
+      listenOnChange: listenOnChange,
+      showNoResultsMessage: _showNoResultsMessage(vis, visData)
     };
-  });
+  }
+
+  render() {
+    const { vis, visData, listenOnChange } = this.props;
+
+    return (
+      <div className="visualization">
+        {this.state.showNoResultsMessage ? (<VisualizationNoResults />) :
+          (<VisualizationChart
+            vis={vis}
+            visData={visData}
+            listenOnChange={listenOnChange}
+          />)
+        }
+      </div>
+    );
+  }
+
+  _onChangeListener = () => {
+    this.forceUpdate();
+  };
+
+  static getDerivedStateFromProps(props, prevState) {
```

Comment:
Could we potentially move the `static` method to top (or less preferred by me: bottom), to not have it sitting somewhere in the middle?

Reply:
i think it makes more sense to order the methods in lifecycle order, but whatever ... this is a real nity pick by the way :)

Reply:
I know. That's why I didn't explicitly state it - I though you'll figure it out anyway :D